Improved support for Go based packages (#14061)
spack/spack: lib/spack/docs/build_systems/gopackage.rst

.. Copyright 2013-2019 Lawrence Livermore National Security, LLC and other
   Spack Project Developers. See the top-level COPYRIGHT file for details.

   SPDX-License-Identifier: (Apache-2.0 OR MIT)

.. _gopackage:

-----------
GoPackage
-----------

There are a couple of things to be aware of about applications built
with Go:

* Go-based projects use a variety of build tools: simple invocations
  of ``go build``, custom go "scripts" that leverage ``go run``, Make,
  and *etc...*.

* Go's dependency management story has evolved from an initial
  do-it-yourself attitude through a small number of commonly used
  tools to an officially supported approach (Go "modules").  The Go
  team provides an `introduction to Go modules
  <https://blog.golang.org/using-go-modules>`_ and maintains a site
  that tracks `the gory details
  <https://github.com/golang/go/wiki/Modules>`_.

  Go modules are the future, ``GOPATH`` will likely be deprecated in
  Go 1.14.

Not surprisingly, the best approach to packaging a Go-based project
for Spack depends on the details of the project.

The ``GoPackage`` base class is intended to make it easy to package

> simple projects.  Tools & techniques that it uses may also be useful

Reviewer: ```suggestion
simple projects.  Tools and techniques that it uses may also be useful
```

when using other build system base classes (e.g. ``MakefilePackage``),
in particular:

- Use the package's ``setup_build_environment`` to set
  ``GO111MODULE=on`` to enable module mode and set

>   ``GOFLAGS="-mod-vendor"`` to enable vendoring, which prevents ``go``

Reviewer: ```suggestion
  ``GOFLAGS=-mod-vendor`` to enable vendoring, which prevents ``go``
```
Quotes shouldn't be needed here, as nothing is ever run through a Bash shell, only a Python shell.

Author: fixed

  them in the package definition via ``import_resources``.

^^^^^^^^^^^^^^^^^^^^^^^^^^
Simple package definitions
^^^^^^^^^^^^^^^^^^^^^^^^^^

The simplest case is a project that uses a simple invocation of ``go
build`` to compile an executable, *and* uses modules, *and* has
vendored its dependencies.  Here's an example of a package definition
which is only slightly complicated by the fact that it requires a
newer version of Go than ``GoPackage`` provides by default:

.. code-block:: python

    from spack import *

    class Lazygit(GoPackage):
        """Simple terminal UI for git commands."""

        homepage = "https://github.com/jesseduffield/lazygit"
        url      = "https://github.com/jesseduffield/lazygit/archive/v0.11.3.tar.gz"

        version('0.11.3', sha256='b3c503de6b34fd4284fd70655e7f42feafc07f090e7f7cc00db261f56c583c46')

        depends_on('go@1.13:', type='build')  # go.mod value overrides default

        executables = ['lazygit']

While it generally follows the common patterns of a Spack package,
it's worth calling attention to its:

* adding a *build* dependency on the version of the Go package that it

>   requires (from the existing `go.mod`);

Reviewer: In other build system docs, I added an entire section on "Important files", where you could describe what `go.mod` is, what to look for, and an example of how it specifies the minimum Go version needed. For example, see https://spack.readthedocs.io/en/latest/build_systems/cmakepackage.html#important-files

Author: I'll improve this.

Author: Interestingly, waf.py, sip.py, and now go.py define and call it as:

```python
def build_args(self):
```

while scons.py and python.py define and call it as:

```python
def build_args(self, spec, prefix):
```

Author: done

* using the default ``build`` and ``install`` steps that ``GoPackage``
  provides; and
* providing the name of the executable to be installed by the default
  ``install`` step.

The ``GoPackage`` provides a `setup_build_environment` that sets

> `GO111MODULES=on` and `GOFLAGS=-mod=vendor` and its default build step

Reviewer: Should these be double backticks?

Author: Oops, markdown slipup....  fixed.

invokes ``go build``.  Additional arguments to ``go build`` can be
provided by setting ``build_args``, e.g. one could

.. code-block:: python

>         build_args = ['-p', '12']  # adjust parallelism of build cmd

Reviewer: I wonder if this shouldn't be a function like in every other build system. Are there any times when you need to access the spec?

Author: That's a better implementation.  I'll update it.

Author: Interestingly, waf.py, sip.py, and now go.py define and call it as:

def build_args(self):
while scons.py and python.py define and call it as:

def build_args(self, spec, prefix):

to invoke ``go build -p 12``.

^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Slightly more complicated packages
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Packages that can be built with a simple invocation of ``go build``,
that use modules, *but have not vendored their dependencies* require a
bit of extra work.  In particular, in order to satisfy Spack's
requirement that nothing is downloaded during the build step, the

> package needs to define "resources" that describe from whence to fetch

Reviewer: Maybe add a link to the Resources section of the Packaging Guide

Author: Tried this

```
:ref:`resources <resources-expanding-extra-tarballs>`
```

but it didn't do what I was hoping for (indeed, it generated an error).

Can I buy a clue?

Author: figured it out....

each dependency, where to emplace it, and the version of the package
to which it applies.  It is possible (likely) that each release of the

> project will new set of resources definitions covering the correct

Reviewer: ```suggestion
project will have a new set of resource definitions covering the correct
```

Author: fixed in a rewrite

versions.

In the general case, compiling these resource is complicated.

Fortunately, packagers of projects that use modules can use an
automated tool (``modules2tuple``) to provide the resource definitions
and ``import_resources`` to include them without overly-complicating
the package definition.  ``modules2tuple``
(https://github.com/dmgk/modules2tuple) was originally developed by
the FreeBSD Ports team to generate the Makefile fragment their build
system uses to represent a Go application's dependencies.  For the
curious, their Porters Handbook includes more details about `packaging
Go Applications
<https://www.freebsd.org/doc/en_US.ISO8859-1/books/porters-handbook/building.html#using-go>`_
for FreeBSD.

There are three steps to describing the dependencies in the package
definition:

> * First, stage the application and use the Go tool chain to

Reviewer: ```suggestion
* First, stage the application and use the Go toolchain to
```

Author: fixed

  solve the set of required dependencies and build a vendor directory.

* Second, run the newly created ``vendor/modules.txt`` file through
  ``module2tuples`` to generate a JSON file containing resource
  definitions.

.. code-block:: console

    # in the root directory of the project (where the go.mod file lives)
    $ go mod vendor

>     $ modules2tuple -spack -app_version=1.2.3 vendor/modules.txt > resources-1.2.3.json

Reviewer: If you indent this whole code block another couple spaces, it should correctly nest with the list

Author: Yep, that's nicer...

>     $ modules2tuple -spack -app_version=1.2.3 vendor/modules.txt > resources-1.2.3.json

Reviewer: Did you add the `-spack` flag to this tool yourself?

Author: Yes, see [the PR](https://github.com/dmgk/modules2tuple/pull/8).  I spent a bunch of time looking at how other projects handle Go applications and FreeBSD's feels cleanest.   Go provides some neat indirection between the name of imports in the code and where it actually needs to be fetched from, and then there's its [revolutionary approach to version selection](https://research.swtch.com/vgo-mvs) that is intended to/seems to/should make long term support of code less painful.  They provide really nice tooling for checksumming and mirroring/proxying and etc..., but it's all Go specific and there aren't any really nice places to hook in other systems that want to handle those bits (aka, us).

There's some awkward stuff (e.g. vanity imports and mirrors) that needs to happen and I didn't want to leave the Spack project having to support a transliteration tool on its own.  Drafting (in the bicycle sense) the FreeBSD team seems easiest.

> * Finally, copy that file into the Spack package directory and add an

Reviewer: I would use `#.` instead of `*` to make a numbered list for these since they're clearly ordered.

Author: Every day, I learn something new....

>   `import_resources` statement that refers to it in the package

Reviewer: ```suggestion
  ``import_resources`` statement that refers to it in the package
```

Author: fixed.

  definition.

.. code-block:: python

        import_resources("resources-1.2.3.py", when="@1.2.3")

^^^^^^^^^^^^^^^
Advanced topics
^^^^^^^^^^^^^^^

Packaging projects that don't fit into either of the previous two

> categories is "left as an exercise for the reader".

Reviewer: 🤣

Author: :smiling_imp:

More seriously, there are two things to keep in mind:

> * the package should use a "run" dependency on go (unless something

Reviewer: Did you mean build here?

Author: yes/yeesh.  Already fixed.

  wacky is happening at run time); and

* you'll need to provide ``resource`` definitions for dependencies
  (``modules2tuples`` might be helpful) and ensure that ``go`` does
  not access the network while building (perhaps by invoking it with
  the ``-mod=vendor`` flag).

* ensure that the use of the ``GOFLAGS`` environment variable does not

>   conflict with attempts to set arguments on the command line.

Reviewer: Can you add a final section with links to more documentation on building Go packages? That's what I did in the other build system docs.

Author: Will do.

Author: done.

<https://github.com/dmgk/modules2tuple>`_ (FreeBSD's tool, which has
been extended to emit Spack resource statements) to provide the
resource definitions and the ``import_resources`` directive to include
them without overly-complicating the package definition.

There are four steps to describing the dependencies in the package
definition:

#. Stage the application and check the ``go.mod`` file for any
   constraints on the go release (before overwriting it as a side
   effect of the next step).

#. Use the Go toolchain and the ``go.mod`` file to determine
   the set of required dependencies and build a vendor directory.

#. Run the newly created ``vendor/modules.txt`` file through
   ``modules2tuple`` to generate a JSON file containing resource
   definitions.

   .. code-block:: console

      # in the root directory of the project (where the go.mod file lives``
      $ go mod vendor
      $ modules2tuple -spack -app_version=1.2.3 vendor/modules.txt > resources-1.2.3.json

#. Finally, copy that file into the Spack package directory and add an
   ``import_resources`` statement that refers to it in the package
   definition.

   .. code-block:: python

>       import_resources("resources-1.2.3.py", when="@1.2.3")

Author: Yep, and there shouldn't be a when clause....

^^^^^^^^^^^^^^^
Advanced topics
^^^^^^^^^^^^^^^

Packaging projects that don't fit into either of the previous two
categories is "left as an exercise for the reader".

More seriously, you'll need to understand how the project builds
itself and use an appropriate Spack build system.  Most projects that
can't use ``GoPackage`` seem to end up using ``MakefilePackage``.

Things to keep in mind include:

* the package should use only a *build* dependency on go (unless

>   something wacky is happening at run time);

Author: Done in the text.  FYI, someone crazy might use `go run` to execute a "go script" or an app might use some bit of the `go` tool, e.g. to format go code via `go fmt`.

   Spack Project Developers. See the top-level COPYRIGHT file for details.

   SPDX-License-Identifier: (Apache-2.0 OR MIT)

.. _gopackage:

-----------
GoPackage
-----------

There are a couple of things to be aware of about applications built
with Go:

* The ``go`` command includes more build/install functionality than
  C's ``cc`` but not as much e.g. Python's ``pip``.

  As a result, Go-based projects use a variety of build strategies:
  simple invocations of ``go build``, custom go "scripts" that
  leverage ``go run``, Makefiles, and/or *etc...*.

* Go's dependency management story has evolved from an initial
  do-it-yourself attitude through a small number of commonly used
  tools to an officially supported approach (Go "modules").  The Go
  team provides an `introduction to Go modules
  <https://blog.golang.org/using-go-modules>`_ and maintains a site
  that tracks `the gory details
  <https://github.com/golang/go/wiki/Modules>`_.

  Go modules are the future, ``GOPATH`` will likely be deprecated in
  Go 1.14.

The best approach to packaging a Go-based project for Spack depends on
the details of the project.  ``GoPackage`` is a good fit for simple

> situations but another package, e.g. ``MakefilePackage``, might be

Reviewer 2: Would it generally be advantageous to combine the logic from `GoPackage` and `MakefilePackage`? Is the issue with doing this the possibility of MRO conflicts?

If yes/yes: I could think about how to combine `GoPackage` with other build systems.

Author: I don't think so, at least immediately.  If something uses a Makefile, than adding dependency on a go compiler, declaring resources via `module2tuples` (or by hand if you're a masochist), and using `MakefilePackage` seems like a nice, simple strategy.  I'm open to something more complicated but don't see any need so far.  At some point down the road it might also make sense to make the go compiler "virtual", satisfiable via `go` or `gccgo` or `tinygo` or one of the LLVM-based compilers.  But not until there's a real use case.

FWIW, here's the current (including this PR) breakdown of build techniques:

| APP           | STRATEGY                  |
|---------------|---------------------------|
| hugo          | go build                  |
| rclone        | go build                  |
| lazygit       | go build                  |
| docui         | go build                  |
| go-md2man     | go build                  |
| lazydocker    | go build                  |
| modules2tuple | go build                  |
| glow          | go build                  |
|               |                           |
| git-lfs       | makefile                  |
| skopeo        | makefile                  |
| umoci         | makefile                  |
| singularity   | makefile                  |
| fzf           | makefile                  |
| direnv        | makefile, w/ weird custom logic |
